github.com/cockroachdb/pebble/internal/metamorphic: TestMetaTwoInstance failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/20971393?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/20971393?buildTab=artifacts#meta) on refs/heads/master @ [ea50351a564c](https://github.com/cockroachdb/pebble/commits/ea50351a564c):


```
          replace_single_delete=true
          threads=2
          disable_value_blocks_for_ingest_sstables=true
          shared_storage_enabled=true
          external_storage_enabled=true
          seed_efos=11008667796612394253
          ingest_split=true
          use_shared_replicate=true
          use_excise=true
          disable_downloads=true
        
        ===== OPS =====
        /artifacts/meta/260114-051815.8893425678956/ops
        ===== HISTORY =====
        /artifacts/meta/260114-051815.8893425678956/random-019/history
        ...[9755 lines skipped]...
        snap45.Close() // <nil> #6138
        // 05:18:21.303 INFO: [JOB 225] sstable deleted 000910
        // 05:18:21.303 INFO: [JOB 225] sstable deleted 000911
        // 05:18:21.303 INFO: [JOB 225] sstable deleted 000912
        // 05:18:21.303 INFO: [JOB 225] sstable deleted 000913
        // 05:18:21.303 INFO: [JOB 225] sstable deleted 000914
        // 05:18:21.303 INFO: [JOB 225] sstable deleted 000915
        // 05:18:21.303 INFO: [JOB 184] flushing: sstable created 000935
        // 05:18:21.303 INFO: [JOB 184] flushing: sstable created 000936
        // 05:18:21.304 INFO: [JOB 184] flushing: sstable created 000937
        // 05:18:21.304 INFO: [JOB 184] flushing: sstable created 000938
        // 05:18:21.304 INFO: [JOB 184] flushing: sstable created 000939
        // 05:18:21.304 INFO: [JOB 225] sstable deleted 000916
        // 05:18:21.304 INFO: [JOB 226] sstable deleted 000549
        // 05:18:21.304 INFO: [JOB 226] sstable deleted 000868
        // 05:18:21.304 INFO: [JOB 226] sstable deleted 000869
        // 05:18:21.304 INFO: [JOB 226] sstable deleted 000909
        db1.Close() // <nil> #6139
        // 05:18:21.305 INFO: [JOB 184] flushing: sstable created 000940
        // 05:18:21.307 INFO: [JOB 184] flushing: sstable created 000941
        // 05:18:21.308 INFO: [JOB 184] flushing: sstable created 000942
        // 05:18:21.308 INFO: [JOB 184] flushing: sstable created 000943
        // 05:18:21.309 INFO: [JOB 184] flushing: sstable created 000944
        // 05:18:21.310 INFO: [JOB 184] flushing: sstable created 000945
        // 05:18:21.310 INFO: [JOB 184] flushing: sstable created 000946
        // 05:18:21.310 INFO: [JOB 184] flushing: sstable created 000947
        // 05:18:21.316 INFO: [JOB 184] flushed 1 memtable (473B) to L0 [000924 000925 000926 000927 000928 000929 000930 000931 000932 000933 000934 000935 000936 000937 000938 000939 000940 000941 000942 000943 000944 000945 000946 000947] (19KB), in 0.0s (0.0s total), output rate 1.7MB/s
        // 05:18:21.316 INFO: [JOB 184] WAL deleted 000901
        db2.Close() // <nil> #6140
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/260114-051815.8893425678956/random-019 --try-to-reduce -v
--- FAIL: TestMetaTwoInstance/execution/random-019 (7.49s)
=== RUN   TestMetaTwoInstance/execution
--- FAIL: TestMetaTwoInstance/execution (0.00s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1768367895889812840 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: PEBBLE-1326

--- a/metamorphic/ops.go
+++ b/metamorphic/ops.go
@@ -2017,6 +2017,7 @@ type replicateOp struct {
 	start, end   UserKey
 }
 
+// runSharedReplicate always closes w.
 func (r *replicateOp) runSharedReplicate(
 	t *Test, h historyRecorder, source, dest *pebble.DB, w *sstable.Writer, sstPath string,
 ) {
@@ -2050,16 +2051,12 @@ func (r *replicateOp) runSharedReplicate(
 			return nil
 		},
 	})
+	err = firstError(err, w.Close())
 	if err != nil {
 		h.Recordf("%s // %v", r.formattedString(t.testOpts.KeyFormat), err)
 		return
 	}
 
-	err = w.Close()
-	if err != nil {
-		h.Recordf("%s // %v", r.formattedString(t.testOpts.KeyFormat), err)
-		return
-	}
 	meta, err := w.Raw().Metadata()
 	if err != nil {
 		h.Recordf("%s // %v", r.formattedString(t.testOpts.KeyFormat), err)
@@ -2084,6 +2081,7 @@ func (r *replicateOp) runSharedReplicate(
 	h.Recordf("%s // %v", r.formattedString(t.testOpts.KeyFormat), err)
 }
 
+// runExternalReplicate always closes w.
 func (r *replicateOp) runExternalReplicate(
 	t *Test, h historyRecorder, source, dest *pebble.DB, w *sstable.Writer, sstPath string,
 ) {
@@ -2122,16 +2120,12 @@ func (r *replicateOp) runExternalReplicate(
 			return nil
 		},
 	})
+	err = firstError(err, w.Close())
 	if err != nil {
 		h.Recordf("%s // %v", r.formattedString(t.testOpts.KeyFormat), err)
 		return
 	}
 
-	err = w.Close()
-	if err != nil {
-		h.Recordf("%s // %v", r.formattedString(t.testOpts.KeyFormat), err)
-		return
-	}
 	meta, err := w.Raw().Metadata()
 	if err != nil {
 		h.Recordf("%s // %v", r.formattedString(t.testOpts.KeyFormat), err)
@@ -2192,10 +2186,12 @@ func (r *replicateOp) run(t *Test, h historyRecorder) {
 	t.opts.Comparer.ValidateKey.MustValidate(r.end)
 	// First, do a RangeKeyDelete and DeleteRange on the whole span.
 	if err := dest.RangeKeyDelete(r.start, r.end, t.writeOpts); err != nil {
+		_ = w.Close()
 		h.Recordf("%s // %v", r.formattedString(t.testOpts.KeyFormat), err)
 		return
 	}
 	if err := dest.DeleteRange(r.start, r.end, t.writeOpts); err != nil {
+		_ = w.Close()
 		h.Recordf("%s // %v", r.formattedString(t.testOpts.KeyFormat), err)
 		return
 	}
@@ -2241,6 +2237,7 @@ func (r *replicateOp) run(t *Test, h historyRecorder) {
 		}
 	}
 	if err := iter.Error(); err != nil {
+		_ = w.Close()
 		h.Recordf("%s // %v", r.formattedString(t.testOpts.KeyFormat), err)
 		return
 	}

